File: exercism/haskell/bob/bob.hs
module Bob where

import Data.Char (isUpper, isAlpha, isSpace)

responseFor :: String -> String
responseFor s
  | isEmpty s = "Fine. Be that way!"
  | isShout s = "Whoa, chill out!"
  | isQuestion s = "Sure."
  | otherwise = "Whatever."
  where
    isEmpty = all isSpace
    isShout s = allUpper && anyAlpha
      where
        alphas = filter isAlpha s
        anyAlpha = length alphas > 0
        allUpper = all isUpper alphas
    isQuestion = (==) '?' . last

File: exercism/clojure/bob/bob.clj
(ns bob
  (use [clojure.string :as str :only [upper-case trim]]))

(defn nothing? [s]
  (empty? (str/trim s)))

(defn yelding? [s]
   (and (= (str/upper-case s) s)
        (re-find #"[A-Z]" s)))

(defn question? [s]
  (= (last (str/trim s)) \?))

(defn response-for [s]
  (cond
   (nothing? s) "Fine. Be that way!"
   (yelding? s) "Woah, chill out!"
   (question? s) "Sure."
   :else "Whatever."))